### PR TITLE
APV data layout refactoring

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
### Purpose of this PR
The goal of this PR is to change the layout of APV data in order to optimize loading and streaming of cells.
With this PR, the data is directly in a format that can be copied to the GPU.
One of the consequence is that all data (all but bricks) contained in the runtime Cell structure are now laid out for GPU which is not the same as the layout of baking cells.
All data needs to be rebaked.
The new data layout should not change visuals at all.
(This PR also renamed debugDisplay structure to probeVolumeDebug because it made no sense :p )

---
### Testing status
- Loaded and baked APV Runtime test
- Tested scenario blending in this scene.
- Tested APV L1/L2 in runtime test scene
- Loaded Sebastien Lachambre APV template and baked it. Tried loading/unloading subscenes.
- Made a big scene to test streaming of data
- Ran a player with this scene.
- Ran graphics tests
- Tested all debug modes for APV

---
### Comments to reviewers
@Unity-Technologies/gfx-qa-hdrp this PR could break a lot of things.
I tried testing as many features as I can but a thorough testing of APV from QA side is a must in this case to make sure I did not miss something.

### CODE REVIEW

**Comment by FrancescoC-unity** [comment]
Comments for future: 

 - Move all the data into GPU layout immediately after we get the data -- so baking cell is not linearly laid out. 
 - Validity neighbour mask should probably be refactored

File: com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs [hunk 1]
```diff
@@ -1065,60 +1140,150 @@ static void WriteBakingCells(ProbeVolumePerSceneData data, List<BakingCell> baki
...
 
                 sceneStateHash = sceneStateHash * 23 + bakingCell.GetBakingHashCode();
 
-                var probesTargetL0L1 = probesL0L1.GetSubArray(startCounts.probesCount * ProbeVolumeAsset.kL0L1ScalarCoefficientsCount, cellCounts.probesCount * ProbeVolumeAsset.kL0L1ScalarCoefficientsCount);
-                for (int j = 0, k = 0; j < cellCounts.probesCount; ++j, k += ProbeVolumeAsset.kL0L1ScalarCoefficientsCount)
-                {
-                    ref var sh = ref bakingCell.sh[j];
-                    WriteToShaderCoeffsL0L1(ref sh, probesTargetL0L1, k);
-                }
+                // Each
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:1200]: Each?

File: com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs [hunk 2]
```diff
@@ -1065,60 +1140,150 @@ static void WriteBakingCells(ProbeVolumePerSceneData data, List<BakingCell> baki
...
                     {
-                        ref var sh = ref bakingCell.sh[j];
-                        WriteToShaderCoeffsL2(ref sh, probesTargetL2, k);
+                        for (int z = 0; z < ProbeBrickPool.kBrickProbeCountPerDim; z++)
+                        {
+                            for (int y = 0; y < ProbeBrickPool.kBrickProbeCountPerDim; y++)
+                            {
+                                for (int x = 0; x < ProbeBrickPool.kBrickProbeCountPerDim; x++)
+                                {
+                                    int index = GetProbeGPUIndex(brickIndex, x, y, z, locSize);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:1241]: Factor out brick loc in a separate fun

File: com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeReferenceVolume.cs [hunk 3]
```diff
@@ -1485,21 +1459,24 @@ internal void Clear()
...
         }
 
+        // Currently only used for 1 chunk at a time but kept in case we need more in the future.
         List<Chunk> GetSourceLocations(int count, int chunkSize)
         {
             var c = new Chunk();
             m_TmpSrcChunks.Clear();
             m_TmpSrcChunks.Add(c);
 
+            var dataLoc = ProbeBrickPool.ProbeCountToDataLocSize(chunkSize);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeReferenceVolume.cs:1469]: Change this to whatever temp data loc is

File: com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs [hunk 4]
```diff
@@ -852,66 +852,135 @@ static void OnLightingDataCleared()
             Clear();
         }
 
-        static void WriteToShaderCoeffsL0L1(ref SphericalHarmonicsL2 sh, NativeArray<float> shaderCoeffsL0L1, int offset)
+        static ushort SHFloatToHalf(float value)
+        {
+            return Mathf.FloatToHalf(value);
```
Comment (adrien-de-tocqueville) [inline com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:857]: does it clamp if value is too big for 16 bit, or does it overflow ?

**Review by iM0ve — CHANGES_REQUESTED** [review]
**Testing status:** In progress

Whats tested:
* APV debug modes
* APV volume overrides
* Rebaking Spotlight Tunnel project and comparing with master
* Rebaking Classroom project and comparing with master
* Virtual offset. Baking with and without.
* Touch up volumes can be created and baked
* Tweaking Probe Volume Settings and rebaking
* Mac and Windows
* Baking with and without Dilation

**Issues:**
1) **FIXED** I can not bake(mac and windows). Getting an error and the probes do not show up. Im using this project. Tried deleting library.
https://github.com/iM0ve/Spotlight-HDRP/tree/APV

Error:
IndexOutOfRangeException: Index was outside the bounds of the array.
UnityEngine.Rendering.ProbeGIBaking.WriteBakingCells (UnityEngine.Rendering.ProbeVolumePerSceneData data, System.Collections.Generic.List`1[T] bakingCells) (at /Users/motiejus.viltrakis/Documents/Git/Graphics/com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:1268)
UnityEngine.Rendering.ProbeGIBaking.OnAdditionalProbesBakeCompleted () (at /Users/motiejus.viltrakis/Documents/Git/Graphics/com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:807)
UnityEditor.Experimental.Lightmapping.Internal_CallAdditionalBakedProbesCompleted () (at /Users/bokken/buildslave/unity/build/Editor/Mono/GI/Lightmapping.bindings.cs:730)

https://user-images.githubusercontent.com/36502659/156151338-ef682db0-d5fe-4fd4-8262-be40fb7ee218.mov

Note: to use the project be sure to checkout APV branch. Open main scene and then add Day.

**Review by iM0ve — APPROVED** [review]
I updated my review. Have not found any more issues.

File: com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs (hunk 4 above)
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:857]: Unity Doc seems to imply it's clamped.

File: com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeBrickIndex.cs [hunk 5]
```diff
@@ -460,11 +464,11 @@ void UpdatePhysicalIndex(Vector3Int brickMin, Vector3Int brickMax, int value, Ce
 
             // Loop through all touched indices
             int chunkStart = cellInfo.firstChunkIndex * kIndexChunkSize;
-            for (int z = brickMin.z; z < brickMax.z; ++z)
+            for (int x = brickMin.x; x < brickMax.x; ++x)
             {
-                for (int y = brickMin.y; y < brickMax.y; ++y)
+                for (int z = brickMin.z; z < brickMax.z; ++z)
```
[the remark this answers is not in the document]
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeBrickIndex.cs:469]: This is done in memory order to optimize memory access.

File: com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs [hunk 6]
```diff
@@ -1137,26 +1302,41 @@ static void WriteBakingCells(ProbeVolumePerSceneData data, List<BakingCell> baki
...
+                    fs.Write(new ReadOnlySpan<byte>(array.GetUnsafeReadOnlyPtr(), array.Length * UnsafeUtility.SizeOf<T>()));
+                    fs.Write(new byte[AlignRemainder16(fs.Position)]);
+                }
+
                 using (var fs = new System.IO.FileStream(cellDataFilename, System.IO.FileMode.Create, System.IO.FileAccess.Write))
-                    fs.Write(new ReadOnlySpan<byte>(probesL0L1.GetUnsafeReadOnlyPtr(), probesL0L1.Length * UnsafeUtility.SizeOf<float>()));
+                {
+                    WriteNativeArray(fs, probesL0L1Rx);
+                    WriteNativeArray(fs, probesL1GL1Ry);
+                    WriteNativeArray(fs, probesL1BL1Rz);
```
[the remark this answers is not in the document]
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.core/Editor/Lighting/ProbeVolume/ProbeGIBaking.cs:1317]: hum yeah nevermind, it's lke 16 bytes max anyaway

File: com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeBrickIndex.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.core/Runtime/Lighting/ProbeVolume/ProbeBrickIndex.cs:469]: yeah but it's accessed that way `int localFlatIdx = z * (size.x * size.y) + x * size.y + y;` so this order is not optimal no ?

**Pull request closed without merging**